ProjFS.Mac/PrjFSKextTests/KextMockUtilities.hpp

```diff
@@ -120,21 +139,40 @@ bool ArgumentsAreEqual(TUPLE_T& arguments, CHECK_TUPLE_T check, std::index_seque
     return (ArgumentIsEqual(std::get<INDICES>(arguments), std::get<INDICES>(check)) && ...);
 }
 
-
 template <typename R, typename... ARGS>
 template <typename... CHECK_ARGS>
-    bool SpecificFunctionCallRecorder<R, ARGS...>::DidCallFunction(FunctionPointerType function, CHECK_ARGS... checkArgs)
+    bool SpecificFunctionCallRecorder<R, ARGS...>::DidCallFunction(FunctionPointerType function, int sequenceNumber, CHECK_ARGS... checkArgs)
 {
     std::pair<typename RecordedCallMap::const_iterator, typename RecordedCallMap::const_iterator> foundCalls = this->recordedCalls.equal_range(function);
     for (typename RecordedCallMap::const_iterator foundCall = foundCalls.first; foundCall != foundCalls.second; ++foundCall)
     {
         if (foundCall->second.argumentValues)
         {
-            if (ArgumentsAreEqual(*foundCall->second.argumentValues, std::forward_as_tuple(checkArgs...), std::index_sequence_for<CHECK_ARGS...>{}))
+            if (ArgumentsAreEqual(*foundCall->second.argumentValues, std::forward_as_tuple(checkArgs...), std::index_sequence_for<CHECK_ARGS...>{}) &&
+                (sequenceNumber == -1 || foundCall->second.callSequenceNumber == sequenceNumber))
             {
                 return true;
             }
         }
     }
     return false;
 }
+
+template <typename R, typename... ARGS>
+template <typename... CHECK_ARGS>
+    bool SpecificFunctionCallRecorder<R, ARGS...>::DidCallFunction(FunctionPointerType function, CHECK_ARGS... checkArgs)
+{
+    return DidCallFunction(function, -1, checkArgs...);
+}
+
+template <typename R, typename... ARGS>
+int SpecificFunctionCallRecorder<R, ARGS...>::CallCount(FunctionPointerType function)
+{
+    int callCount = 0;
+    std::pair<typename RecordedCallMap::const_iterator, typename RecordedCallMap::const_iterator> foundCalls = this->recordedCalls.equal_range(function);
```

Comment:
I know some of my code has declarations like this as well, but I'm thinking it's less readable than `auto` at this point. We don't have to go for `auto`, there are a few in-between options which I think would also be preferable over the status quo:

 * I think the latest C++ standard allows partial type declarations, so in this case, I believe `std::pair foundCalls = …;` should work. This at least makes it obvious we're getting a `pair` - and pretty much all operations on STL containers return iterators in some form, so not spelling those out in detail seems sensible.
 * Locally Typedef'ing the iterator type to something much shorter.

Reply:
> I think the latest C++ standard allows partial type declarations, so in this case, I believe std::pair foundCalls = …; should work. This at least makes it obvious we're getting a pair - and pretty much all operations on STL containers return iterators in some form, so not spelling those out in detail seems sensible.

Cool I did not know about this feature 👍 

> I know some of my code has declarations like this as well, but I'm thinking it's less readable than auto at this point. 

I agree on this point, and template programming is one spot where things can get messy without `auto`.  

If the above approach does not work out I'm good with using `auto` here with some descriptive variable names.